exploration-module/src/main/java/io/edpn/backend/exploration/infrastructure/kafka/sender/RequestDataMessageKafkaSenderService.java

```diff
@@ -25,28 +24,18 @@ public class RequestDataMessageKafkaSenderService implements KafkaSenderService<
     private final KafkaTopicHandler kafkaTopicHandler;
     private final KafkaTemplate<String, JsonNode> jsonNodekafkaTemplate;
 
-    private final Semaphore semaphore = new Semaphore(1);
-
     @Override
-    public void sendToKafka(RequestDataMessage requestDataMessage) {
-        try {
-            semaphore.acquire();
-            if (requestDataMessageRepository.find(requestDataMessage).isEmpty()) {
-                requestDataMessageRepository.create(requestDataMessage);
-            } else {
-                log.debug("info request already queued");
-            }
-
-            sendPendingMessages();
-        } catch (InterruptedException e) {
-            Thread.currentThread().interrupt();
-            log.error("Failed to acquire semaphore", e);
-        } finally {
-            semaphore.release();
+    public synchronized void sendToKafka(RequestDataMessage requestDataMessage) {
+        if (requestDataMessageRepository.find(requestDataMessage).isEmpty()) {
+            requestDataMessageRepository.create(requestDataMessage);
+        } else {
+            log.debug("info request already queued");
         }
+
+        sendPendingMessages();
     }
 
-    private void sendPendingMessages() {
+    private synchronized void sendPendingMessages() {
         requestDataMessageRepository.findNotSend()
                 .forEach(requestDataMessage ->
```

Comment:
Doesn't have to do with this PR in particular, but... "findNotSend"? Doesn't sound proper to my ear, perhaps "findUnsent" or something similar?

Also, unless `sendPendingMessages()` is invoked from elsewhere, it doesn't _need_ to be synchronized, since `sendToKafka()` already acquires the lock - OTOH doesn't really hurt, just in case it will be called from elsewhere. Since the lock is already acquired, it's uncontested here and doesn't really affect performance.

Got a bit curious about the synchronization here and I'm wondering... is it really necessary and/or doing what it should? What internal state of _this_ class is being guarded by synchronization? If the repository is accessed elsewhere (as in not through these methods), synchronization here is largely meaningless, and should happen in the repository instead. Also, for preventing a check-then-act kind of race condition, it would perhaps be better if the repository offered an atomic "send unsent" operation. :thinking:

Reply:
If the purpose of `synchronized` here is exclusiveness (ie. no calls "let in" while one is in progress) for calls to this service, and for that, this would IMHO work just fine.